### PR TITLE
2/3 Remove the need for an HTTP server

### PR DESCRIPTION
**Pull request by mickael-menu**
## Review notes

The Settings API PRs are numbered in the order of merge, but should be reviewed in reverse order, starting with https://github.com/readium/kotlin-toolkit/pull/141.

This PR refactor the EPUB navigator to support serving the resources without an HTTP server. This was necessary to inject the new user settings API.

### How resources are served?

The `WebViewClient`'s `shouldInterceptRequest()` API is used to intercept and serve the resources under the custom origin `https://readium`.

* Publication resources are served manually under the base URL `https://readium/publication/`. It supports byte range requests for media resources.
* Files in the app's `assets/` directory are served under the base URL `https://readium/assets/` using `WebViewAssetLoader`.
    * The custom assets must be explicitly allowed to be served, using `EpubNavigatorFragment.Configuration.servedAssets`. All assets under the `readium/` folder are automatically allowed.

## How to opt-in?

The `EpubNavigatorFragment` is using this new method of serving the EPUB resources by default, unless you provide the `baseUrl` argument, which is now optional.

:warning: Using this new method is incompatible with the legacy EPUB settings API.

### Error page

The PR introduces a (overridable in the test app) error page for when a resource fails to be loaded. Previously the navigator was broken and you were stuck in the default web view 404 page.

---

* Fixes https://github.com/readium/kotlin-toolkit/issues/34

### CODE REVIEW

File: readium/shared/src/main/java/org/readium/r2/shared/fetcher/Resource.kt [hunk 1]
```diff
@@ -210,6 +211,47 @@ class FailureResource(private val link: Link, private val error: Resource.Except
             
 }
 
+/**
+ * Resource that will act as a proxy to the first given resource that is not a failure.
+ */
+class FallbackResource(private val originalResource: Resource, private vararg val fallbackResources: Resource) : Resource {
```
[the remark this answers is not in the document]
Comment (mickael-menu) [reply readium/shared/src/main/java/org/readium/r2/shared/fetcher/Resource.kt:217]: Writing this, I realize that we should probably have a 500 page when the failure error is not "NotFound".

File: test-app/src/main/assets/readium/404.xhtml [hunk 2]
```diff
@@ -0,0 +1,13 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<!DOCTYPE html PUBLIC "-//W3C//DTD XHTML 1.1//EN"
+    "http://www.w3.org/TR/xhtml11/DTD/xhtml11.dtd">
+
+<html xmlns="http://www.w3.org/1999/xhtml" xml:lang="en">
+<head>
+    <title>Resource not found</title>
```
Comment (mickael-menu) [inline test-app/src/main/assets/readium/404.xhtml:7]: It's exactly the same as the `404.xhtml` in the navigator module. I copied it as an example on how to override it in the test app.

File: readium/shared/src/main/java/org/readium/r2/shared/fetcher/Resource.kt [hunk 3]
```diff
@@ -24,6 +24,7 @@ import org.readium.r2.shared.extensions.requireLengthFitInt
 import org.readium.r2.shared.parser.xml.ElementNode
 import org.readium.r2.shared.parser.xml.XmlParser
 import org.readium.r2.shared.publication.Link
+import org.readium.r2.shared.publication.services.cover
```
Comment (qnga) [inline readium/shared/src/main/java/org/readium/r2/shared/fetcher/Resource.kt:27]: What the fuck is that? There are a few unused imports through the code base, that one is the worst one I could see.

File: readium/shared/src/main/java/org/readium/r2/shared/fetcher/Resource.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (qnga) [reply readium/shared/src/main/java/org/readium/r2/shared/fetcher/Resource.kt:217]: A specific error page would make sense  for `Forbidden` too. We should probably match `Resource.Exception` cases.

File: readium/shared/src/main/java/org/readium/r2/shared/fetcher/Resource.kt (hunk 3 above)
Comment (stevenzeck) [reply readium/shared/src/main/java/org/readium/r2/shared/fetcher/Resource.kt:27]: Yeah a right-click of the readium directory and Optimize Imports should always be done.  Wonder if there is a Gradle task or checker.

Comment (mickael-menu) [reply readium/shared/src/main/java/org/readium/r2/shared/fetcher/Resource.kt:27]: I'm not sure where it came from either, [the commit that introduced it doesn't use it](https://github.com/readium/kotlin-toolkit/commit/07d58d32cc969c9e4704331e96a57a4871c3f249#diff-9f256285c5d2609af688ffd63eff2b7990c47301bc7422a20a823fd87354166b). Maybe Android Studio auto-importing something with a failed auto-completion?

I didn't know about this global optimize imports, that's helpful thanks @stevenzeck. I'll do this for this PR but it's about time we setup an auto-formatter like `ktlint` which tackles that automatically.

File: readium/shared/src/main/java/org/readium/r2/shared/fetcher/Resource.kt (hunk 1 above)
[the remark this answers is not in the document]
Comment (mickael-menu) [reply readium/shared/src/main/java/org/readium/r2/shared/fetcher/Resource.kt:217]: Yep

File: test-app/src/main/java/org/readium/r2/testapp/reader/EpubReaderFragment.kt [hunk 4]
```diff
@@ -280,7 +280,7 @@ private fun annotationMarkTemplate(context: Context, @ColorInt defaultTint: Int
                 width: 30px;
                 height: 30px;
                 border-radius: 50%;
-                background: url('$imageUrl') no-repeat center;
+                background: url('https://readium/assets/annotation-icon.svg') no-repeat center;
```
[the remark this answers is not in the document]
Comment (mickael-menu) [reply test-app/src/main/java/org/readium/r2/testapp/reader/EpubReaderFragment.kt:283]: Good idea

[the remark this answers is not in the document]
Comment (qnga) [reply test-app/src/main/java/org/readium/r2/testapp/reader/EpubReaderFragment.kt:283]: Sounds good!

**Pull request merged**